File: lazyc/lazyc/compiler/ast_print.ml
// lazyc/compiler/ast_print.ml
//
// Print AST nodes in the same text format as src/ast_print.c, so that
// AST dumps can be diffed across compilers for cross-validation.
//
// Step 21e: expression printing only. Statements come in 21f.

// Print n levels of indentation (2 spaces each).
Long ast_print_indent(Long d) {
    Long i = 0;
    while (i < d) {
        print("  ");
        i = i + 1;
    }
    return 0;
}

// Print a Ptr<Type> as the type name. Mirrors src/ast_print.c::type_name
// — note this prints the SHORT form ("Ptr" not "Ptr<...>") to match.
Long ast_print_type_short(Ptr<Type> t) {
    if (t == cast<Ptr<Type>>(null)) {
        print("?");
        return 0;
    }
    Long k = (*t).kind;
    if (k == TY_BOOLEAN())  { print("Boolean");  return 0; }
    if (k == TY_CHAR())     { print("Char");     return 0; }
    if (k == TY_BYTE())     { print("Byte");     return 0; }
    if (k == TY_INTEGER())  { print("Integer");  return 0; }
    if (k == TY_UINTEGER()) { print("uInteger"); return 0; }
    if (k == TY_WHOLE())    { print("Whole");    return 0; }
    if (k == TY_UWHOLE())   { print("uWhole");   return 0; }
    if (k == TY_LONG())     { print("Long");     return 0; }
    if (k == TY_ULONG())    { print("uLong");    return 0; }
    if (k == TY_STRING())   { print("String");   return 0; }
    if (k == TY_PTR())      { print("Ptr");      return 0; }
    if (k == TY_STRUCT())   { print("struct");   return 0; }
    if (k == TY_VOID())     { print("void");     return 0; }
    print("?");
    return 0;
}

// If e->ety is set (typechecker has run), print " :TypeName". Otherwise
// nothing. Mirrors C ast_print.c::ty_suffix exactly.
Long ast_ty_suffix(Ptr<Expr> e) {
    Ptr<Type> t = (*e).ety;
    if (t == cast<Ptr<Type>>(null)) { return 0; }
    if ((*t).kind == TY_UNKNOWN()) { return 0; }
    print(" :");
    ast_print_type_short(t);
    return 0;
}

// Print one expression node and its children, recursively, indented.
Long ast_print_expr(Ptr<Expr> e, Long d) {
    ast_print_indent(d);
    Long k = (*e).kind;

    if (k == EX_NUMBER()) {
        print("Number ");
        // We need to print a Long without %l format... use println? Actually
        // we have buf_push_long for buffers, but here we're streaming to
        // stdout. Use println but suppress its newline by streaming pieces.
        // Easiest: build into a Buf, then write_bytes. But we don't have
        // raw write_bytes exposed. Use println to emit the number alone.
        // The C compiler prints "Number %lld" then ty_suffix then '\n'.
        // We approximate with println("%l", ...) which adds a newline, so
        // we need a no-newline version. lazyc has only print/println; the
        // print function does NOT add a newline. Use it.
        print("%l", (*e).num);
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_CHAR_LIT()) {
        print("Char '%c'", cast<Char>((*e).char_val));
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_STRING_LIT()) {
        // C uses printf("%.*s") to print the slice without the
        // surrounding quotes added by lazyc. For us, str_data is the
        // raw inner content already.
        print("String \"%s\"", cast<String>((*e).str_data));
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_BOOL_LIT()) {
        if ((*e).bool_val == 1) { print("Bool true"); }
        else                    { print("Bool false"); }
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_NULL()) {
        print("Null");
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_IDENT()) {
        print("Ident %s", cast<String>((*e).name));
        ast_ty_suffix(e);
        println("");
        return 0;
    }
    if (k == EX_BINARY()) {
        print("Binary %s", cast<String>(op_name((*e).op)));
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        ast_print_expr((*e).child1, d + 1);
        return 0;
    }
    if (k == EX_UNARY()) {
        print("Unary %s", cast<String>(op_name((*e).op)));
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        return 0;
    }
    if (k == EX_CALL()) {
        print("Call %s", cast<String>((*e).name));
        ast_ty_suffix(e);
        println("");
        Ptr<PtrVec> args = (*e).call_args;
        if (args != cast<Ptr<PtrVec>>(null)) {
            Long n = (*args).count;
            Long i = 0;
            while (i < n) {
                Ptr<Expr> a = cast<Ptr<Expr>>(ptrvec_get(args, i));
                ast_print_expr(a, d + 1);
                i = i + 1;
            }
        }
        return 0;
    }
    if (k == EX_CAST()) {
        print("Cast<");
        ast_print_type_short((*e).cast_target);
        print(">");
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        return 0;
    }
    if (k == EX_ADDR_OF()) {
        print("AddrOf");
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        return 0;
    }
    if (k == EX_DEREF()) {
        print("Deref");
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        return 0;
    }
    if (k == EX_FIELD()) {
        print("Field .%s", cast<String>((*e).name));
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        return 0;
    }
    if (k == EX_INDEX()) {
        print("Index");
        ast_ty_suffix(e);
        println("");
        ast_print_expr((*e).child0, d + 1);
        ast_print_expr((*e).child1, d + 1);
        return 0;
    }
    println("?Expr kind=%l", k);
    return 0;
}

File: lazyc/src/ast_print.c
#include "ast.h"
#include <stdio.h>

static const char *type_name(TypeKind k) {
    switch (k) {
        case TY_BOOLEAN:  return "Boolean";
        case TY_CHAR:     return "Char";
        case TY_BYTE:     return "Byte";
        case TY_INTEGER:  return "Integer";
        case TY_UINTEGER: return "uInteger";
        case TY_WHOLE:    return "Whole";
        case TY_UWHOLE:   return "uWhole";
        case TY_LONG:     return "Long";
        case TY_ULONG:    return "uLong";
        case TY_STRING:   return "String";
        case TY_PTR:      return "Ptr";
        case TY_STRUCT:   return "struct";
        case TY_VOID:     return "void";
        default:          return "?";
    }
}
static const char *op_name(OpKind o) {
    static const char *n[] = {"+","-","*","/","%","==","!=","<",">","<=",">=","neg","!"};
    return n[o];
}
static void indent(int d) { for (int i=0;i<d;i++) fputs("  ", stdout); }

static void ty_suffix(Expr *e) {
    if (e->type.kind != TY_UNKNOWN) printf(" :%s", type_name(e->type.kind));
}

static void print_expr(Expr *e, int d) {
    indent(d);
    switch (e->kind) {
        case EX_NUMBER:     printf("Number %lld", e->num); ty_suffix(e); putchar('\n'); break;
        case EX_CHAR_LIT:   printf("Char '%c'", e->ch); ty_suffix(e); putchar('\n'); break;
        case EX_STRING_LIT: printf("String \"%.*s\"", (int)e->str.len, e->str.data); ty_suffix(e); putchar('\n'); break;
        case EX_BOOL_LIT:   printf("Bool %s", e->boolean?"true":"false"); ty_suffix(e); putchar('\n'); break;
        case EX_NULL:       printf("Null"); ty_suffix(e); putchar('\n'); break;
        case EX_IDENT:      printf("Ident %.*s", (int)e->ident.len, e->ident.name); ty_suffix(e); putchar('\n'); break;
        case EX_BINARY:
            printf("Binary %s", op_name(e->bin.op)); ty_suffix(e); putchar('\n');
            print_expr(e->bin.lhs, d+1);
            print_expr(e->bin.rhs, d+1);
            break;
        case EX_UNARY:
            printf("Unary %s", op_name(e->un.op)); ty_suffix(e); putchar('\n');
            print_expr(e->un.operand, d+1);
            break;
        case EX_CALL:
            printf("Call %.*s", (int)e->call.name_len, e->call.name); ty_suffix(e); putchar('\n');
            for (size_t i=0;i<e->call.nargs;i++) print_expr(e->call.args[i], d+1);
            break;
        case EX_CAST:
            printf("Cast<%s>", type_name(e->cast.target.kind)); ty_suffix(e); putchar('\n');
            print_expr(e->cast.operand, d+1);
            break;
        case EX_ADDR_OF:
            printf("AddrOf"); ty_suffix(e); putchar('\n');
            print_expr(e->addr.target, d+1);
            break;
        case EX_DEREF:
            printf("Deref"); ty_suffix(e); putchar('\n');
            print_expr(e->deref.operand, d+1);
            break;
        case EX_FIELD:
            printf("Field .%.*s", (int)e->field.name_len, e->field.name);
            ty_suffix(e); putchar('\n');
            print_expr(e->field.operand, d+1);
            break;
        case EX_INDEX:
            printf("Index"); ty_suffix(e); putchar('\n');
            print_expr(e->index.base, d+1);
            print_expr(e->index.index, d+1);
            break;
    }
}

static void print_stmt(Stmt *s, int d) {
    indent(d);
    switch (s->kind) {
        case ST_VAR_DECL:
            printf("VarDecl %s %.*s\n", type_name(s->var.ty.kind),
                   (int)s->var.name_len, s->var.name);
            if (s->var.init) print_expr(s->var.init, d+1);
            break;
        case ST_ASSIGN:
            printf("Assign %.*s\n", (int)s->assign.name_len, s->assign.name);
            print_expr(s->assign.value, d+1);
            break;
        case ST_PTR_STORE:
            printf("PtrStore\n");
            print_expr(s->ptr_store.target, d+1);
            print_expr(s->ptr_store.value, d+1);
            break;
        case ST_FIELD_STORE:
            printf("FieldStore\n");
            print_expr(s->field_store.target, d+1);
            print_expr(s->field_store.value, d+1);
            break;
        case ST_INDEX_STORE:
            printf("IndexStore\n");
            print_expr(s->index_store.target, d+1);
            print_expr(s->index_store.value, d+1);
            break;
        case ST_IF:
            printf("If\n");
            print_expr(s->if_s.cond, d+1);
            indent(d); printf("Then:\n"); print_stmt(s->if_s.then_b, d+1);
            if (s->if_s.else_b) { indent(d); printf("Else:\n"); print_stmt(s->if_s.else_b, d+1); }
            break;
        case ST_WHILE:
            printf("While\n");
            print_expr(s->while_s.cond, d+1);
            print_stmt(s->while_s.body, d+1);
            break;
        case ST_FOR:
            printf("For\n");
            if (s->for_s.init)   { indent(d+1); printf("Init:\n");   print_stmt(s->for_s.init, d+2); }
            if (s->for_s.cond)   { indent(d+1); printf("Cond:\n");   print_expr(s->for_s.cond, d+2); }
            if (s->for_s.update) { indent(d+1); printf("Update:\n"); print_stmt(s->for_s.update, d+2); }
            print_stmt(s->for_s.body, d+1);
            break;
        case ST_RETURN:
            printf("Return\n");
            if (s->ret.value) print_expr(s->ret.value, d+1);
            break;
        case ST_BREAK:
            printf("Break\n");
            break;
        case ST_CONTINUE:
            printf("Continue\n");
            break;
        case ST_BLOCK:
            printf("Block\n");
            for (size_t i=0;i<s->block.n;i++) print_stmt(s->block.stmts[i], d+1);
            break;
        case ST_EXPR:
            printf("ExprStmt\n");
            print_expr(s->expr_s.expr, d+1);
            break;
    }
}

void print_program(Program *p) {
    for (size_t i=0; i<p->nfuncs; i++) {
        FuncDecl *f = p->funcs[i];
        if (f->is_extern) {
            printf("Extern %s %.*s(", type_name(f->return_ty.kind),
                   (int)f->name_len, f->name);
        } else {
            printf("Func %s %.*s(", type_name(f->return_ty.kind),
                   (int)f->name_len, f->name);
        }
        for (size_t j=0;j<f->nparams;j++) {
            if (j) printf(", ");
            printf("%s %.*s", type_name(f->params[j].ty.kind),
                   (int)f->params[j].name_len, f->params[j].name);
        }
        printf(")\n");
        if (!f->is_extern) print_stmt(f->body, 1);
    }
}
